runtime/src/comm/ugni/comm-ugni-launch.c

```diff
@@ -23,16 +23,72 @@
 
 #define _POSIX_C_SOURCE 200112L  // for setenv(3) in <stdlib.h>
 
+#include <assert.h>
+#include <math.h>
 #include <stdlib.h>
 
 #include "chplrt.h"
+#include "comm-ugni-heap-pages.h"
 #include "chpl-comm-launch.h"
 #include "error.h"
 
 
+//
+// With mem=jemalloc and comm=ugni with hugepages and a dynamically
+// extensible heap, we have to configure jemalloc's large chunk size.
+// It would be preferable to deal with this entirely within the target
+// program, but we need it before jemalloc does its constructor-based
+// initialization and we don't have a dependable way to arrange that
+// in the target program.
+//
+static
+void maybe_set_jemalloc_lg_chunk(void) {
+  size_t hps;
+
+  hps = chpl_comm_ugni_getHeapPageSize();
+  if (hps == chpl_comm_ugni_getSysPageSize())
+    return;
+
+  //
+  // We're using the jemalloc memory layer, the heap is dynamically
+  // extensible, and we've got hugepages, so we'll be registering
+  // memory.  (chpl_comm_ugni_getHeapPageSize()) checks all this.)
+  //
+  // Arrange for jemalloc's chunk size (and alignment) to be the same
+  // as what the comm layer recommends, which is the heap page size.
+  // Make sure this is a power of 2, because jemalloc needs that.
+  //
+  int hps_log2;
+  char* ev;
+  char buf[1000];
+
+  // sanity check: power of 2 and >= sys page size
+  assert((hps & (hps - 1)) == 0
+         && hps >= chpl_comm_ugni_getSysPageSize());
+
+  // lg_chunk is specified as the base-2 log of the expansion chunk size
+  hps_log2 = lrint(log2((double) hps));
+
+  if ((ev = getenv("CHPL_JE_MALLOC_CONF")) == NULL) {
```

Comment:
We shouldn't be hardcoding `CHPL_JE_MALLOC_CONF`. We should be using an uppercased `CHPL_JEMALLOC_PREFIX`, which is defined in third-party/jemalloc/Makefile.include

If that isn't defined then we may be using a system jemalloc (not officially supported today, but may be in the future) so we should default to `JE_`. 

The top of `runtime/include/mem/jemalloc/chpl-mem-impl.h` shows what we do with getting the prefix for the alloc routines

Reply:
Also, I'm trying to figure out the best way to maintain this code if for example a jemalloc upgrade changes the name of the env var. I'm not going to remember that this code needs to be updated, so I was wondering if we could put a function in the jemalloc shim or header for this code to call. 

I'm not sure how that would work since this will be called from the launcher though.. 

At the very least I think we need a note in the jemalloc README about checking this file during version upgrades, but i'm open to other suggestions

Reply:
Fixed the jemalloc README in 88c499a6.